feat: Add MCP server

nzakas opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

-   [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[x] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This change adds an MCP server that is triggered by `--mcp`. Because no other options are valid with this, I implemented it similar to `--init`.

The TypeScript SDK is a bit rough, so I had to work around a couple of things. Overall, everything works. You can test it out locally using the MCP inspector:
https://github.com/modelcontextprotocol/inspector

fixes #19588

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

nzakas commented:
I've got this working and it's kind of fun. :smile: Also added setup docs.

nzakas commented on docs/src/use/mcp.md:
```diff
@@ -0,0 +1,82 @@
+
+## Setting Up ESLint MCP Server in VS Code
+
+To add the ESLint MCP server to your VS Code workspace, follow these steps:
```
Are you referring to the Copilot extension?

nzakas commented on docs/src/use/mcp.md:
```diff
@@ -0,0 +1,82 @@
+3. Toggle on the ESLint MCP server tools using the "Tools" button in the chat view
+4. Ask Copilot to perform ESLint tasks, such as:
+    - "Check this file for linting errors"
+    - "Fix all ESLint issues in the current file"
```
Ooh, that is interesting. I'm not sure what the expected behavior there should be. I'll play around a bit.

fasttime replied:
Maybe we could say "Lint and fix the current file with ESLint". With that prompt, Copilot indicates when a problem isn't autofixable and asks before fixing it.

![image](https://github.com/user-attachments/assets/92db6c02-b4ad-4aae-ae26-e08bfc0a9044)

nzakas replied:
It's running the same tool in both cases, so I'm not sure that wording is necessarily deterministic. I think there may be some way to provide more instructions for the tool. I'll look into it.

nzakas replied:
I've tried changing the name of the tool a few different times to see if that helps, but it doesn't. Maybe we just don't provide a tool for fixing as it seems Copilot will just go and try to fix things on its own?

fasttime replied:
Yes, the outcome is not deterministic. Sometimes Copilot doesn't even use the MCP server tool and instead it suggests to run `npx eslint --fix` directly.

nzakas commented on bin/eslint.js:
```diff
@@ -155,6 +155,26 @@ ${getErrorMessage(error)}`;
+
+		process.on("SIGINT", () => {
+			mcpServer.close();
+			console.log("HI");
```
Hehe, definitely a mistake.

nzakas commented:
Okay, I removed the "lint and fix" tool. I also removed the prompts as it's unclear how this is intended to be used and I'd rather start small and add things later as we find their helpful.

fasttime commented on lib/mcp/mcp-server.js:
```diff
@@ -0,0 +1,55 @@
+	"Lint files using ESLint. This will not modify the files and will report any issues found.",
+	filePathsSchema,
+	async ({ filePaths }) => {
+		const eslint = new ESLint({
+			// enable lookup from file rather than from cwd
+			flags: ["unstable_config_lookup_from_file"],
+		});
+
+		const results = await eslint.lintFiles(filePaths);
```
In VS Code (I haven't checked other IDEs yet) the MCP server runs with cwd set to my user's home folder. This is reflected in the `cwd` constructor option and it could affect how plugins work when files are linted. Are we comfortable with this? Alternatively we could set the current directory explicitly with `process.chdir` to avoid surprises.

nzakas replied:
I'm not sure there's anything we can do about this. I don't think we have a logical directory to pass to `process.chdr()`.

As a note, this is why we need the `unstable_config_lookup_from_file` flag passed to correctly find a config file for any files being linted.

fasttime replied:
I thought that maybe we could set the current directory explicitly to `os.homedir()` or to `/` in case the server is started from a different location.

nzakas replied:
I this calling `process.chdir()` is a bit dangerous. It's possible that they're still trying to figure out what the `cwd` for MCP servers and I don't think we want to override that default.

nzakas replied:
It looks like Cursor sets the cwd to the location of the Cursor executable. I'd say let's hold off on trying to change the directory. Maybe there's a way we can get the editor to give us a relevant directory.

fasttime replied:
Windsurf also passes its own cwd to the MCP server process. We can start with the editor default then.

fasttime commented on docs/src/use/mcp.md:
```diff
@@ -0,0 +1,152 @@
+Once configured, you can use ESLint tools with Cascade by asking it to:
+
+- Check files for linting errors
+- Fix ESLint issues in specific files
```
Shall we keep this line also without the prompt to fix files?

```suggestion
```

caridy commented on lib/mcp/mcp-server.js:
```diff
@@ -0,0 +1,55 @@
+});
+
+const filePathsSchema = {
+	filePaths: z.array(z.string()).nonempty(),
```
try to be a lot more explicit here, otherwise the superagent/orchestrator/mcp client might give you the wrong path, something like this:

```suggestion
	filePaths: z.array(z.string().min(1).describe('Full filesystem path a file to be analyzed')).nonempty(),
```

I suspect some evolution is going to be needed around this at the MCP level.

caridy commented on lib/mcp/mcp-server.js:
```diff
@@ -0,0 +1,55 @@
+		return {
+			content: results.map(result => ({
+				type: "text",
+				text: JSON.stringify(result),
```
@nzakas this is where I think we have to put more effort. Most MCP servers will just return the result of the operation (calling into some tools or apis) without decorating that with some natural language explanation. In my experience, when you decorate it with some prefix and suffix in natural language, providing concrete instructions on what just happened and what to do next, that helps the super-agent to understand better the output, and the intent. Obviously, eslint is everywhere, and so the LLMs have a pretty good idea of what this output means, but I think some little instructions will help to control the LLM instead of being aggressive toward resolving all issues. Something like "... You must ask for confirmation before attempting to correct the issues identified above" can go a long way for the super-agent.

nzakas replied:
Ah, gotcha. Thanks so much for the tip.

nzakas commented:
After updating the tool output with explicit instructions I'm no longer having the agent automatically fix the errors. Thanks @caridy!

fasttime commented on lib/mcp/mcp-server.js:
```diff
@@ -0,0 +1,72 @@
+	version: pkg.version,
+});
+
+const filePathsSchema = {
+	filePaths: z
+		.array(
+			z
+				.string()
+				.min(1)
+				.describe("Full filesystem path of a file to be analyzed"),
+		)
+		.nonempty(),
+};
```
I'm trying to run the tool in Cursor and it's telling me that `filePaths` has a wrong type. I guess this is because something isn't right with the schema?

![image](https://github.com/user-attachments/assets/620d453f-769b-4b46-9eff-e6a76aac073a)

My server config in `.cursor/mcp.json` looks like this:

```json
{
	"mcpServers": {
		"eslint": {
			"command": "node",
			"args": ["/Users/fasttime/eslint/bin/eslint.js", "--mcp"],
			"env": {}
		}
	}
}
```

nzakas replied:
Okay, so it doesn't like the `describe()` in the file paths.

fasttime reviewed (approved):
LGTM, thanks! This is good enough for a start. I left some comments but I don't have any further suggestions at this time. I would like another review from a team member, if possible.

fasttime commented:
Oh, and we should probably add `.temp-eslintsuppressions` and `.cursor/` to gitignore.

nzakas commented:
I opted to fix the test so that `.temp-eslintsuppressions` is deleted after completion.

nzakas commented:
Ping @mdjermanovic.

christian-kreuzberger-dtx commented:
While I think this change is awesome to have, it just introduced a ton of new dependencies for `eslint`.

Please consider reverting this PR and invest into #19682 .

RobinTail commented:
@nzakas 

If one doesn't use AI with ESLint, why do they have to install all that?
Could you make it an opt-in feature, an optional peer dependency?

![image](https://github.com/user-attachments/assets/7853740b-6ee9-44f4-8eae-0da2e575c8e9)

SimenB commented:
There is a PR open for extracting it: #19699

RobinTail commented:
@SimenB , please release Jest 30
